aws/lambda/log-classifier/src/rule_match.rs

```diff
@@ -19,17 +22,38 @@ pub struct SerializedMatch {
     line: String,
     line_num: usize,
     captures: Vec<String>,
+    /// The optional context where this failure occurs. This is a free-form
+    /// stack of strings that includes the last commands before the failure
+    pub context: Vec<String>,
 }
 
 impl SerializedMatch {
     pub fn new(m: &Match, log: &Log) -> SerializedMatch {
         // Unwrap because we know this is a valid key (since the Log object is never mutated.)
         let line = log.lines.get(&m.line_number).unwrap();
+
+        let mut context = Vec::with_capacity(CONTEXT_DEPTH);
+        // NB: backtrack the log till we find the previous command. This relies
+        // on GitHub console log convention to prefix a bash command with +. An
+        // important note is that this only works for bash though, not Windows
+        // PowerShell. But Windows support could be added later.
+        for i in (1..=m.line_number).rev() {
+            if context.len() == CONTEXT_DEPTH {
+                break
+            }
+
+            let l = log.lines.get(&i).unwrap();
+            if l.starts_with("+") {
```

Comment:
lgtm, though I wonder if it's possible to expand this to just expand the context window generally.

Reply:
Let's increase the depth to a higher value, then we will have data to see what a good value is.  Ideally, I think gathering all commands executed by the failed step is a good option.  However, I find it a bit tricky to implement as there is no separator on the log itself.  I wonder how GitHub could separate the log into multiple steps on their service.

Reply:
I make the size of the context window a parameter with a default value, so it could be set by the client